File: foialens/lib/tools/extract-entities.ts
import Anthropic from '@anthropic-ai/sdk';
import pool from '@/lib/db/client';
import type { EntityEntry, EntityType } from '@/lib/types';
import { extractText, parseJSON } from './haiku-utils';

const anthropic = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY });

const HAIKU = 'claude-haiku-4-5-20251001';
const MAX_CHUNKS = 20;

export type EntityScope = string;  // 'full' or a document ID

type ExtractedEntity = Omit<EntityEntry, 'firstSeenRunId'>;

export async function extractEntities(
  scope: EntityScope,
  workspaceId: string,
  knownNames: Set<string> = new Set(),
): Promise<{ entities: ExtractedEntity[]; newCount: number }> {
  const chunks = await fetchChunks(scope, workspaceId);
  if (chunks.length === 0) return { entities: [], newCount: 0 };

  const context = chunks
    .map(c => `[p.${c.start_page}] ${c.content}`)
    .join('\n\n---\n\n');

  const response = await anthropic.messages.create({
    model: HAIKU,
    max_tokens: 2048,
    system:
      'You are a precise information extraction system. ' +
      'Return ONLY valid JSON with no prose, preamble, or markdown fences.',
    messages: [
      {
        role: 'user',
        content: `Extract every named entity from the document text below.

Return a JSON array where each element has exactly these fields:
- "name": entity name (string)
- "type": one of "person" | "organization" | "date" | "amount" | "location"
- "mentions": approximate count of times mentioned (number)
- "pageRefs": page numbers where found (number[])
- "representativeContext": one sentence showing the entity in context (string)

Document text:
${context}

Return ONLY the JSON array.`,
      },
    ],
  });

  const text = extractText(response);
  const parsed = parseJSON<ExtractedEntity[]>(text);
  if (!parsed || !Array.isArray(parsed)) return { entities: [], newCount: 0 };

  const entities = parsed.filter(isValidEntity);
  const newCount = entities.filter(
    e => !knownNames.has(e.name.toLowerCase()),
  ).length;

  return { entities, newCount };
}

// ── Helpers ───────────────────────────────────────────────────────────────────

async function fetchChunks(scope: EntityScope, workspaceId: string) {
  if (scope === 'full') {
    const { rows } = await pool.query<{ content: string; start_page: number }>(
      `SELECT content, start_page FROM chunks
       WHERE workspace_id = $1
       ORDER BY document_id, chunk_index
       LIMIT $2`,
      [workspaceId, MAX_CHUNKS],
    );
    return rows;
  }

  // scope is a document ID
  const { rows } = await pool.query<{ content: string; start_page: number }>(
    `SELECT content, start_page FROM chunks
     WHERE workspace_id = $1 AND document_id = $2
     ORDER BY chunk_index
     LIMIT $3`,
    [workspaceId, scope, MAX_CHUNKS],
  );
  return rows;
}

const VALID_TYPES = new Set<EntityType>([
  'person', 'organization', 'date', 'amount', 'location',
]);

function isValidEntity(e: unknown): e is ExtractedEntity {
  if (!e || typeof e !== 'object') return false;
  const obj = e as Record<string, unknown>;
  return (
    typeof obj.name === 'string' &&
    VALID_TYPES.has(obj.type as EntityType) &&
    typeof obj.mentions === 'number' &&
    Array.isArray(obj.pageRefs) &&
    typeof obj.representativeContext === 'string'
  );
}

File: foialens/backend/tools/extract_entities.py
import asyncio

from db.client import pool
from .haiku_utils import HAIKU, call_with_backoff, extract_text, parse_json
from .search_documents import search_documents

MAX_CHUNKS = 50
VALID_TYPES = {"person", "organization", "location"}

# Queries designed to surface entity-dense chunks across different entity types.
_ENTITY_QUERIES = [
    "signed by director secretary chief officer commissioner",
    "department agency bureau division office city county state",
    "contractor vendor company corporation LLC incorporated",
    "appointed hired promoted transferred reassigned resigned",
    "located address headquarters district region jurisdiction",
]


async def extract_entities(
    scope: str,
    workspace_id: str,
    known_names: set[str] | None = None,
) -> dict:
    if known_names is None:
        known_names = set()

    chunks = await _fetch_chunks(scope, workspace_id)
    if not chunks:
        return {"entities": [], "newCount": 0}

    context = "\n\n---\n\n".join(f"[p.{r['start_page']}] {r['content']}" for r in chunks)

    _SCHEMA = {
        "type": "json_schema",
        "json_schema": {
            "name": "entity_extraction",
            "strict": True,
            "schema": {
                "type": "object",
                "properties": {
                    "entities": {
                        "type": "array",
                        "items": {
                            "type": "object",
                            "properties": {
                                "name":                  {"type": "string"},
                                "type":                  {"type": "string", "enum": ["person", "organization", "location"]},
                                "mentions":              {"type": "number"},
                                "pageRefs":              {"type": "array", "items": {"type": "number"}},
                                "representativeContext": {"type": "string"},
                            },
                            "required": ["name", "type", "mentions", "pageRefs", "representativeContext"],
                            "additionalProperties": False,
                        },
                    }
                },
                "required": ["entities"],
                "additionalProperties": False,
            },
        },
    }
    response = await call_with_backoff(
        model=HAIKU,
        max_tokens=8192,
        response_format=_SCHEMA,
        messages=[
            {
                "role": "system",
                "content": (
                    "You are a data extraction API for legal and investigative research. "
                    "Extract named persons, organizations, and locations. "
                    "Focus on named officials and staff, government agencies, contractors, vendors, and named places."
                ),
            },
            {
                "role": "user",
                "content": f"Document text:\n{context}",
            },
        ],
    )

    raw_text = extract_text(response)
    print(f"[extract_entities] raw response ({len(raw_text)} chars): {raw_text[:300]!r}", flush=True)
    parsed = parse_json(raw_text)
    if isinstance(parsed, dict):
        parsed = parsed.get("entities", [])
    if not isinstance(parsed, list):
        print(f"[extract_entities] parse_json failed; raw={raw_text[:300]!r}", flush=True)
        return {"entities": [], "newCount": 0}

    print(f"[extract_entities] parsed {len(parsed)} items, {sum(1 for e in parsed if _is_valid(e))} valid", flush=True)
    entities = [e for e in parsed if _is_valid(e)]
    new_count = sum(1 for e in entities if e["name"].lower() not in known_names)

    return {"entities": entities, "newCount": new_count}


async def _fetch_chunks(scope: str, workspace_id: str) -> list:
    if scope != "full":
        rows = await pool().fetch(
            "SELECT content, start_page FROM chunks "
            "WHERE workspace_id = $1 AND document_id = $2 ORDER BY chunk_index LIMIT $3",
            workspace_id, scope, MAX_CHUNKS,
        )
        print(f"[extract_entities] single-doc fetch: {len(rows)} chunks", flush=True)
        return rows

    db_count = await pool().fetchval("SELECT count(*)::int FROM chunks WHERE workspace_id = $1", workspace_id)
    print(f"[extract_entities] workspace_id={workspace_id!r} db_chunk_count={db_count}", flush=True)

    results = await asyncio.gather(
        *[search_documents(q, workspace_id, limit=12) for q in _ENTITY_QUERIES]
    )

    seen: set[str] = set()
    chunks: list[dict] = []
    for q, result in zip(_ENTITY_QUERIES, results):
        hits = result.get("results", [])
        print(f"[extract_entities] query={q!r:.50} → {len(hits)} hits", flush=True)
        for r in hits:
            if r["chunkId"] not in seen:
                seen.add(r["chunkId"])
                chunks.append({"start_page": r["startPage"], "content": r["content"]})

    print(f"[extract_entities] total unique chunks: {len(chunks)}", flush=True)
    return chunks[:MAX_CHUNKS]


def _is_valid(e: object) -> bool:
    if not isinstance(e, dict):
        return False
    return (
        isinstance(e.get("name"), str)
        and e.get("type") in VALID_TYPES
        and isinstance(e.get("mentions"), (int, float))
        and isinstance(e.get("pageRefs"), list)
        and isinstance(e.get("representativeContext"), str)
    )
